### ISSUE TITLE
append_if_no_line inserts a duplicate line if a line exists with a trailing whitespace

### ISSUE THREAD

**Issue by jasonwbarnett** [sous-chefs/line#90]
Given the following resource:

```
append_if_no_line 'my resource' do
  path '/tmp/some_file'
  line 'this is my line'
end
```

It will append the line, even if the file contains the following content (trailing whitespace):

```
this is my line 
```

**Comment by jasonwbarnett** [comment]
Question: Is this the intended behavior? If not, I'll commit a fix for this.

**Comment by MarkGibbons** [comment]
Hi @jasonwbarnett, 
  That's intended behavior.  Trailing whitespace may be significant.
   You could groom the file with a replace_or_add resource that replaces 'this is my line\s+' with 'this is my line' or use replace_or_add to append the line.
Mark

**Comment by jasonwbarnett** [comment]
I have an idea. What if we add two properties:

`ignore_trailing_whitespace` and `ignore_leading_whitespace` and have both default to `false` which is the current behavior.

This would allow individuals to describe intent explicitly.

Thoughts?

**Comment by josephholsten** [comment]
I understand the desire for this, but it then opens the question: how would this interact with `pattern`?

Because it potentially opens up a large set of edge cases I don’t want to make desicions about, I’m opposed to the proposed properties.

Instead, I’d rather we added some doc to the readme. Perhaps just add the above as examples.

**Comment by jasonwbarnett** [comment]
@josephholsten I'm confused by your response. The `append_if_no_line` resource does not have a `pattern` property. What do you mean, "how would this interact with `pattern`?"

**Comment by josephholsten** [comment]
Oh my, I misread! Yes, this sounds like a bug. If you could write a test and fix, that would be awesome. In the event I’m wrong, a concrete test case would still be useful.

**Comment by jasonwbarnett** [comment]
I’ll write some unit tests and the code itself and submit a pull later this week, at the latest.

Thanks

**Comment by MarkGibbons** [comment]
@jasonwbarnett 
Are the leading and trailing white spaces a common issue for you?  Quite a few edge cases would get introduced with this change so there should be some pretty good reason for doing it.

Just making the changes for append is pretty simple. Add \s+ at the start or end of the pattern check regex as appropriate.

I looked at the resources, replace_or_add would need the same type of change, some what harder though, to retain consistency.  Processing trailing spaces in the list management resources is pretty arbitrary and should be cleaned up to match. The delete lines resource doesn't care. 

I'm also trying to figure out how this would interact with multi line inserts which is another outstanding request. There are also some other ideas floating around for a more general processing pattern.  I was going to use some of the time at chef-conf to work through some possibilities.

thanks,
Mark

**Comment by jasonwbarnett** [comment]
I just started writing test cases and I'll submit what I have now. I think that `replace_or_add` is fundamentally different because it's based on a `Regexp` versus the "line as a whole" concept. I don't know that this is an extremely common use case, but it was an unexpected behavior and so I thought I would contribute something more explicit over implicit (you could easily argue that could be fixed with docs instead -- open to that as an option).

**Comment by MarkGibbons** [comment]
A key piece of replace_or_add is matching the inserted line. It also matches a pattern.

**Comment by jasonwbarnett** [comment]
@MarkGibbons The current design comes across as treating them differently.

1. The two receive different arguments (`String` vs `Regexp`)
2. One converts a string to a `Regexp` and the other literally assumes the `Regexp` begins with `^` and ends with `$` and it escapes the string.

I could see this being solved by either the pull I've already submitted or allowing someone to pass a `Regexp` directly, like the other resource.

**Comment by MarkGibbons** [comment]
You may be overlooking this line in replace or add.
````
if line =~ regex || line == new_resource.line
````
We do check for the regex, we also look for an explicit match to the line, white space and all.

**Comment by mengesb** [comment]
So just for giggles; say for instance this modified a python script, then spacing becomes critically important, so matching on exactly that spacing is the difference between success and failure. I concur with @MarkGibbons that this is explicitly intended behavior.

I'm wondering if we're trying to creatively solve the "spurious space" winding up in a variable which should be solved by cleaner entry or functions like `chomp()` instead of trying to over-engineer and handle fuzzy logic.

**Comment by josephholsten** [comment]
Can we compare specific use cases with example code using a modified `append_if_no_line` vs current `replace_or_add`? We’re clearly talking about something that will require an example in the README, so let’s describe how we’re going to explain this users.

@jasonwbarnett What’s this look like using existing resource? 

```ruby
append_if_no_line 'my resource' do
  path '/tmp/some_file'
  line 'this is my line'
  ignore_trailing_whitespace true
  ignore_leading_whitespace true
end
```

Is there a case this solves that fundamentally couldn’t be converted?

**Comment by MarkGibbons** [comment]
@mengesb Thank you. Sounds like a test case to be added.

@josephholsten Good point. replace_or_add may already be able to do most of this.  Jason's issue was that the behavior of append was surprising.  That may be a documentation problem. I think your example above becomes something like:
````
replace_or_add  'my resource' do
  path '/tmp/some_file'
  line 'this is my line'
  pattern /\s+this is my line\s+/
  append_only  true # New option needed otherwise '    this is my line' would be replaced and @mengesb 's comment about python becomes a real thing.
end
````

@jasonwbarnett I do want to thank you for identifying an issue and coming up with a PR so fast.  I personally don't think the behavior is surprising.  We are talking about an exact line append.
Not matching 'this is my line\b\b\b' is the same to me as not matching 'this is my\b\b\b\b line'.  The extra white space makes them different lines.  Other folks, other opinions though.

 If we want to introduce the capability of appending a line only if something close to the line isn't found then I think adding an append_only option to replace_or_add would be pretty efficient.  Would you be willing to update some documentation (README.md) to clear up the confusion and add an option & tests to replace_or_add to allow for the behavior you suggest?  You should pull the master branch down if you do.  We're prepping to release version 2 of line very shortly.


Thanks,
   Mark

**Comment by MarkGibbons** [comment]
Closed by documentation update in #104.
